### PR TITLE
chore: bump to datafusion 39, arrow 52, pyo3 0.21

### PR DESCRIPTION
**Pull request by abhiaagarwal**
# Description
Updates the arrow and datafusion dependencies to 52 and 39(-rc1) respectively. This is necessary for updating pyo3.

While most changes with trivial, some required big rewrites. Namely, the logic for the Updates operation had to be rewritten (and simplified) to accommodate some new sanity checks inside datafusion: (https://github.com/apache/datafusion/pull/10088).

Depends on delta-kernel having its arrow and object-store version bumped as well. This PR doesn't include any major changes for pyo3, I'll open a separate PR depending on this PR.

# Related Issue(s)

# Documentation

### CODE REVIEW

File: crates/core/src/operations/update.rs [hunk 1]
```diff
@@ -307,66 +304,24 @@ async fn execute(
         },
     ));
 
-    // Perform another projection but instead calculate updated values based on
-    // the predicate value.  If the predicate is true then evalute the user
-    // provided expression otherwise return the original column value
-    //
-    // For each update column a new column with a name of __delta_rs_ + `original name` is created
```
Comment (abhiaagarwal) [inline crates/core/src/operations/update.rs:314]: I would really appreciate if someone could take a look at this diff — I think it should replicate the exact logic except heavily simplified and it passes all tests, but there could be an edge case I'm missing.

**Comment by abhiaagarwal** [comment]
This is now ready for review. All python and rust tests are passing locally on my machine (docs seem to have issues building, though). This only enables the gil-refs feature in the python bindings, it does not migrate to the new API (will be done in a follow-up PR)

File: crates/core/src/delta_datafusion/expr.rs [hunk 2]
```diff
@@ -276,33 +272,6 @@ impl<'a> Display for SqlFormat<'a> {
                     write!(f, "{expr} IN ({})", expr_vec_fmt!(list))
                 }
             }
-            Expr::GetIndexedField(GetIndexedField { expr, field }) => match field {
```
Comment (ion-elgreco) [inline crates/core/src/delta_datafusion/expr.rs:279]: Why is this being removed?

File: crates/core/src/operations/update.rs (hunk 1 above)
Comment (ion-elgreco) [reply crates/core/src/operations/update.rs:314]: Looks fine to me the change, before it was creating these __delta_rs columns but seems unnecessary to first add those columns and then select only those. Doing this in one projection seems much better

File: crates/core/src/delta_datafusion/expr.rs (hunk 2 above)
Comment (abhiaagarwal) [reply crates/core/src/delta_datafusion/expr.rs:279]: Those functions don't exist before, and based on my understanding from https://github.com/apache/datafusion/pull/10568, those rewrites happen earlier now so the functionality stays the same, it's just done by DF

File: crates/core/src/operations/update.rs (hunk 1 above)
Comment (abhiaagarwal) [reply crates/core/src/operations/update.rs:314]: Believe me when I say that I stared at this for a couple hours before figuring out what it was doing 🥳 should probably give a nice speed increase too, I didn't benchmark the old code but I did benchmark the projection operation in general and it's very expensive

File: crates/core/src/delta_datafusion/expr.rs (hunk 2 above)
Comment (ion-elgreco) [reply crates/core/src/delta_datafusion/expr.rs:279]: Just wondering if we can catch the new functions and keep the old display formatting

File: python/src/filesystem.rs [hunk 3]
```diff
@@ -590,31 +580,23 @@ impl ObjectOutputStream {
 
     fn write(&mut self, data: &PyBytes) -> PyResult<i64> {
         self.check_closed()?;
-        let len = data.as_bytes().len() as i64;
         let py = data.py();
-        let data = data.as_bytes();
-        let res = py.allow_threads(|| match rt().block_on(self.writer.write_all(data)) {
```
Comment (ion-elgreco) [inline python/src/filesystem.rs:596]: Did you forget re-adding this? Now the GIL will be in place

File: python/src/filesystem.rs [hunk 4]
```diff
@@ -590,31 +580,23 @@ impl ObjectOutputStream {
...
-            }
-        })?;
-        self.buffer_size += len;
-        if self.buffer_size >= self.max_buffer_size {
-            let _ = self.flush(py);
-            self.buffer_size = 0;
+        let bytes = data.as_bytes();
+        self.buffer.extend_from_slice(bytes);
+        let len = bytes.len();
+        if self.max_buffer_size >= self.buffer.content_length() {
```
Comment (ion-elgreco) [inline python/src/filesystem.rs:587]: This should be the other way around, right?

To only flush when the current buffer is greater than the max_buffer_size (this was added a while ago to prevent deadlocks)

File: crates/core/src/delta_datafusion/expr.rs (hunk 2 above)
Comment (abhiaagarwal) [reply crates/core/src/delta_datafusion/expr.rs:279]: Hmmmm, I'm not entirely sure this is possible. Those operations have been converted into ScalarFunctions, so the concept of "taking an index of a field" is now "a scalar function that operates on a field and an index". Had to modify a bunch of tests to accommodate. https://github.com/delta-io/delta-rs/blob/af5997cf7dd8cce97b07b5c8c78e23cb7092549f/crates/core/src/delta_datafusion/expr.rs#L656-L659

It could be possible manually match on scalar functions and convert them into the old style, though.

File: python/src/filesystem.rs (hunk 3 above)
Comment (abhiaagarwal) [reply python/src/filesystem.rs:596]: Yep, but the write function isn't async or multithreaded (it just extends a buffer in memory) so I don't believe releasing the GIL has any impact there? It releases it in `flush` if it gets called.

File: python/src/filesystem.rs (hunk 4 above)
Comment (abhiaagarwal) [reply python/src/filesystem.rs:587]: Haha yep! Explains that one odd test failure, thank you for catching.

File: python/src/filesystem.rs (hunk 3 above)
Comment (ion-elgreco) [reply python/src/filesystem.rs:596]: > Yep, but the write function isn't async or multithreaded (it just extends a buffer in memory) so I don't believe releasing the GIL has any impact there? It releases it in `flush` if it gets called. 

I think it's more that some folks want to run stuff in parallel/threaded and just to release the GIL whenever possible, doesn't hurt to keep it that way with allow_threads

Comment (abhiaagarwal) [reply python/src/filesystem.rs:596]: added it back! also figured I'd add a check in `close` to make sure we're only flushing if the buffer has data inside it.

File: crates/core/src/delta_datafusion/expr.rs (hunk 2 above)
Comment (ion-elgreco) [reply crates/core/src/delta_datafusion/expr.rs:279]: Hmm, it would make the roundtrippable format closer to SQL, but I guess with the timestamp parsing I already kind of stopped doing that 😆

**Review by ion-elgreco — APPROVED** [review]

**Comment by ion-elgreco** [comment]
Nice work! @abhiaagarwal

**Comment by abhiaagarwal** [comment]
Thank you @ion-elgreco! DF 39 should be released later this week, and the upstream delta-kernel needs to be resolved as well. I'll open a follow-up PR for moving to the new pyo3 lifetimes, and I'm also interested in implementing an async-native API using [`pyo3-asyncio`](https://github.com/awestlake87/pyo3-asyncio) since a good chunk of my work code can benefit from it :)

**Comment by ion-elgreco** [comment]
@abhiaagarwal yeah leaving it unmerged for now. Also checking with rest on slack if they are fine with keeping the git reference of delta-kernel-rs for time being.

Great, you can ping me when you got the new bound APIs working!

Just two things regarding pyo3-asyncio. It seems the original maintainer is not very active anymore, so it's still locked to pyo3 0.20 unfortunately. 

The other thing is, where do you need it for ;) as most of the async stuff is internal

**Comment by abhiaagarwal** [comment]
@ion-elgreco yeah, but it appears that the maintainer of pyo3 is interested in taking stewardship over it, so maybe that'll be resolved soon.

I'm mostly interested to see if there are any perf improvements. Most of the code I use with the python deltalake package in already is async native (ie. async main), having async methods won't "infect" the rest of our codebase. We use it as a pseudo backend with FastAPI, for reference. There appears to be work being done in the main pyo3 repo that'll allow the gil to be released in async methods, so we could have the best of both worlds.

**Comment by ion-elgreco** [comment]
@abhiaagarwal can you bump delta-kernel-rs to 0.1.1, it includes your changes now. And also bump the python cargo toml to 0.18.1?

**Comment by abhiaagarwal** [comment]
@ion-elgreco sure, I'll do it by EOD. Also will bump DF since they released 39

File: Cargo.toml [hunk 5]
```diff
@@ -26,33 +26,33 @@ debug = true
 debug = "line-tables-only"
 
 [workspace.dependencies]
-delta_kernel = { version = "0.1" }
+delta_kernel = { git = "https://github.com/abhiaagarwal/delta-kernel-rs", branch = "arrow_52" }
```
Comment (rtyler) [inline Cargo.toml:29]: this should now be updated to 0.1.1

**Comment by abhiaagarwal** [comment]
Bumped to published versions, also added a new test that seemingly confirms that the issues related to decimals in #1778 et al on both engines seem to be fixed

**Comment by ion-elgreco** [comment]
> Bumped to published versions, also added a new test that seemingly confirms that the issues related to decimals in #1778 et al on both engines seem to be fixed 

Yes I was waiting for some time for the arrow bump since they added Scientific notation support for decimals :)

Thanks for adding the test! 🙂

**Review by ion-elgreco — APPROVED** [review]
